### PR TITLE
feat: implement Files view "empty" states - WPB-19388

### PR DESCRIPTION
**Pull request by jullianm**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-19388" title="WPB-19388" target="_blank">StoryWPB-19388</a>  [iOS] As a user I want to see an empty state if there are no files
  </td></table>
  <br />
 

### Issue

This PR introduces states handling in the Files components (view and view model):
- loading : show progress view while fetching nodes
- received: show nodes
- noData: show "empty" data view
- pending: show "files are being prepared" view

it also introduces a `SyncCellsStateUseCase` to ensure the `cellsState` value is up to date before opening the Files view in case the cells were previously in a pending state and might have been updated since then (this should be an edge case though).

### Testing

_Describe how to test._

_Optional: attachments like images, videos, etc._

---

### Checklist

- [ ] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [ ] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [ ] Make sure you use the API for UI elements that support large fonts.
- [ ] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [ ] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewItemView.swift [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import SwiftUI
+import WireDesign
+
+struct FilesViewItemView: View {
```
Comment (jullianm) [inline WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewItemView.swift:25]: put it in own file

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesView.swift [hunk 2]
```diff
@@ -23,8 +23,8 @@ import WireMessagingDomain
 import WireMessagingDomainSupport
 import WireReusableUIComponents
 
-private typealias Strings = L10n.Localizable.Conversation.WireCells
-private typealias Accessibility = L10n.Accessibility.Conversation.WireCells
+typealias Strings = L10n.Localizable.Conversation.WireCells
+typealias Accessibility = L10n.Accessibility.Conversation.WireCells
 
 package struct FilesView: View {
```
Comment (jullianm) [inline WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesView.swift]: many changes in this file, mostly due to splitting views in smaller pieces

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewItemView.swift (hunk 1 above)
Comment (samwyndham) [reply WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewItemView.swift:25]: Haha I've done the same in https://github.com/wireapp/wire-ios/pull/3511. Merge will be fun 🥳

File: WireMessaging/Tests/WireMessagingTests/Resources/SnapshotResults/FilesViewTests/testFilesViewItemView_withLongStrings.light.png
Comment (samwyndham) [inline WireMessaging/Tests/WireMessagingTests/Resources/SnapshotResults/FilesViewTests/testFilesViewItemView_withLongStrings.light.png:1]: question/issue: I'm confused about this. Didn't this already exist?

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/ConversationViewController.swift [hunk 3]
```diff
@@ -279,6 +284,7 @@ final class ConversationViewController: UIViewController {
 
         resolveConversationIfOneOnOne()
         updateVerificationStatusIfNeeded()
+        syncCellsStateIfPending()
```
Comment (samwyndham) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/ConversationViewController.swift]: suggestion: Move to view did appear just so that it is triggered in more cases.

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesView.swift [hunk 4]
```diff
@@ -23,8 +23,8 @@ import WireMessagingDomain
 import WireMessagingDomainSupport
 import WireReusableUIComponents
 
-private typealias Strings = L10n.Localizable.Conversation.WireCells
-private typealias Accessibility = L10n.Accessibility.Conversation.WireCells
+typealias Strings = L10n.Localizable.Conversation.WireCells
+typealias Accessibility = L10n.Accessibility.Conversation.WireCells
```
Comment (samwyndham) [inline WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesView.swift]: issue: keep this private and duplicate wherever else it is needed. Otherwise these values are usable throughout WireMessagingUI event though they are specific to WireCells.

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift [hunk 5]
```diff
@@ -52,16 +52,37 @@ package final class FilesViewModel: ObservableObject {
         static let loadMoreThreshold = 5
     }
 
+    enum State {
```
Comment (samwyndham) [inline WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift]: suggestion: Make `Equatable`. This is really useful to have and removes the need of the manual conformance in tests.

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift [hunk 6]
```diff
@@ -52,16 +52,37 @@ package final class FilesViewModel: ObservableObject {
...
     private let fetchNodesUseCase: WireCellsFetchNodesUseCase
+    private let isCellsStatePending: Bool
 
-    package init(fetchNodesUseCase: WireCellsFetchNodesUseCase) {
+    package init(
+        fetchNodesUseCase: WireCellsFetchNodesUseCase,
+        isCellsStatePending: Bool
+    ) {
         self.fetchNodesUseCase = fetchNodesUseCase
+        self.isCellsStatePending = isCellsStatePending
```
Comment (samwyndham) [inline WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift]:
```suggestion
        self.State = isCellsStatePending ? .pending : .loading
```

I think this works just as well and allows us to delete the `isCellsStatePending` property.

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift [hunk 7]
```diff
@@ -52,16 +52,37 @@ package final class FilesViewModel: ObservableObject {
...
     }
 
-    @Published private(set) var items: [FilesViewItem] = []
     @Published private var nextPageToken: WireCellsPageToken?
     @Published private var loadMoreTask: LoadItemsTask?
     @Published var alert: AlertModel?
+    @Published var state: State = .loading
```
Comment (samwyndham) [inline WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift]: praise: Replacing items with state to limit possible states.

**Review by samwyndham — CHANGES_REQUESTED** [review]
Nice work! I added a few small suggestions.

File: WireMessaging/Tests/WireMessagingTests/Resources/SnapshotResults/FilesViewTests/testFilesViewItemView_withLongStrings.light.png
Comment (jullianm) [reply WireMessaging/Tests/WireMessagingTests/Resources/SnapshotResults/FilesViewTests/testFilesViewItemView_withLongStrings.light.png:1]: it was added in SnapshotResults, not sure why it shows up here. Do we need that SnapshotResults folder at all ? I thought ReferenceImages was sufficient.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/ConversationViewController.swift (hunk 3 above)
Comment (jullianm) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/ConversationViewController.swift]: done in https://github.com/wireapp/wire-ios/pull/3512/commits/9bd117a84545ab1d1a0d82405e94487df553b06f thx

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesView.swift (hunk 4 above)
Comment (jullianm) [reply WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesView.swift]: done in https://github.com/wireapp/wire-ios/pull/3512/commits/9bd117a84545ab1d1a0d82405e94487df553b06f thx

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift (hunk 5 above)
Comment (jullianm) [reply WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift]: 👍  done in https://github.com/wireapp/wire-ios/pull/3512/commits/9bd117a84545ab1d1a0d82405e94487df553b06f

File: WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift (hunk 6 above)
Comment (jullianm) [reply WireMessaging/Sources/WireMessagingUI/WireCells/Components/Files/FilesViewModel.swift]: done in https://github.com/wireapp/wire-ios/pull/3512/commits/9bd117a84545ab1d1a0d82405e94487df553b06f

**Review by samwyndham — APPROVED** [review]

File: WireMessaging/Tests/WireMessagingTests/Resources/SnapshotResults/FilesViewTests/testFilesViewItemView_withLongStrings.light.png
Comment (samwyndham) [reply WireMessaging/Tests/WireMessagingTests/Resources/SnapshotResults/FilesViewTests/testFilesViewItemView_withLongStrings.light.png:1]: I'm not sure how it got there but those files shouldn't be tracked. I've hopefully untracked it now.

**Review by johnxnguyen — APPROVED** [review]
